src/basic/os-util.c

```diff
@@ -67,6 +70,56 @@ int open_extension_release(const char *root, const char *extension, char **ret_p
                 r = chase_symlinks(extension_full_path, root, CHASE_PREFIX_ROOT,
                                    ret_path ? &q : NULL,
                                    ret_fd ? &fd : NULL);
+                /* Cannot find the expected extension-release file? The image filename might have been
+                 * mangled on deployment, so fallback to checking for any file in the extension-release.d
+                 * directory, and return the first one with a user.extension-release xattr instead.
+                 * The xattr is checked to ensure the author of the image considers it OK if names do
+                 * not match. */
+                if (r == -ENOENT) {
+                        _cleanup_free_ char *extension_release_dir_path = NULL;
+                        _cleanup_closedir_ DIR *extension_release_dir = NULL;
+
+                        r = chase_symlinks_and_opendir("/usr/lib/extension-release.d/", root, CHASE_PREFIX_ROOT,
+                                                       &extension_release_dir_path, &extension_release_dir);
+                        if (r < 0)
+                                return r;
+
+                        r = -ENOENT;
+                        struct dirent *de;
+                        FOREACH_DIRENT(de, extension_release_dir, return -errno) {
+                                if (de->d_type != DT_REG)
+                                        continue;
+
+                                if (!startswith(de->d_name, "extension-release"))
+                                        continue;
+
+                                if (!image_name_is_valid(de->d_name))
+                                        continue;
+
+                                _cleanup_free_ char *p = NULL;
+                                p = path_join(extension_release_dir_path, de->d_name);
+                                if (!p)
+                                        return -ENOMEM;
+
+                                if (getxattr(p, "user.extension-release", NULL, 0) < 0)
+                                        continue;
+
+                                r = 0; /* Found it! */
+
+                                if (ret_fd) {
+                                        /* We already chased the directory, and checked that
+                                         * this is a real file and not a symlink, so just open it. */
+                                        fd = open(p, O_RDONLY|O_CLOEXEC|O_PATH);
```

Comment:
maybe open the file via openat() first (which is what fgetxattrat_fake() does internally), and then use that fd for both the xattr checing and for returning here.